csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -163,50 +164,126 @@ public static void WriteBool(ref Span<byte> buffer, ref WriterInternalState stat
                     {
-                        fixed (char* sourceChars = &MemoryMarshal.GetReference(source))
-                        fixed (byte* destinationBytes = &MemoryMarshal.GetReference(buffer.Slice(state.position)))
-                        {
-                            bytesUsed = Utf8Encoding.GetBytes(sourceChars, source.Length, destinationBytes, buffer.Length);
-                        }
+                        bytesUsed = Encoding.UTF8.GetBytes(sourceChars, source.Length, destinationBytes, destinationBuffer.Length);
                     }
-                    state.position += bytesUsed;
-#endif
                 }
+
+                state.position += bytesUsed;
+#endif
             }
-            else
+        }
+
+        private static void WriteStringMultiBuffer(ref Span<byte> buffer, ref WriterInternalState state, string value)
+        {
+#if NETSTANDARD1_1
+            // slowpath when Encoding.GetBytes(Char*, Int32, Byte*, Int32) is not available
+            byte[] bytes = Utf8Encoding.GetBytes(value);
+            WriteRawBytes(ref buffer, ref state, bytes);
+#else
+            if (state.CodedOutputStream != null)
             {
-                // Opportunity for future optimization:
-                // Large strings that don't fit into the current buffer segment
-                // can probably be optimized by using Utf8Encoding.GetEncoder()
-                // but more benchmarks would need to be added as evidence.
+                // When writing to Encoder we need at least 4 bytes available in the buffer to ensure
+                // there is space to write a character. Because CodedOutputStream always writes to the
+                // end of its buffer we can't guarantee that so use slow path.
                 byte[] bytes = Utf8Encoding.GetBytes(value);
                 WriteRawBytes(ref buffer, ref state, bytes);
+                return;
+            }
+
+            // The destination byte array is not be large enough so multiple writes are required.
+            // Encoder will keep state of unwritten data.
+            if (state.stringEncoder == null)
+            {
+                state.stringEncoder = Encoding.UTF8.GetEncoder();
+            }
+
+            ReadOnlySpan<char> source = value.AsSpan();
+            bool completed = false;
+
+            // When completed is true then all chars from the source string have been consumed and
+            // the encoder has no state left to write as bytes.
+            while (!completed)
+            {
+                // Refresh the buffer with a minimum size of the maximum unicode char byte size.
+                // A minimum buffer size is required because at least one unicode character must
+                // be written when Encoder.Convert is called. 
+                WriteBufferHelper.RefreshBuffer(ref buffer, ref state, sizeHint: MaximumBytesPerUtf8Char);
```

Comment:
I'm not a fan of introducing the need for having a sizeHint in refreshbuffer for just this specific corner case.

It sounds like having  less than MaximumBytesPerUtf8Char bytes left would be a pretty rare condition, so perhaps if that happens, we could write those bytes into a stackalloc'd buffer and then write the resulting bytes into the destination using WriteRawBytes(span)?